lib/srv/db/postgres/test.go

```diff
@@ -364,6 +371,106 @@ func (s *TestServer) handleCreateStoredProcedure(query string) error {
 	return nil
 }
 
+// multiMessage wraps *pgproto3.DataRow and implements pgproto3.BackendMessage by writing multiple copies of this message in Encode.
+type multiMessage struct {
+	singleMessage *pgproto3.DataRow
+	payload       []byte
+}
+
+func newMultiMessage(rowSize, repeats int) (*multiMessage, error) {
+	buf := make([]byte, rowSize)
+	_, err := rand.Read(buf)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	message := &pgproto3.DataRow{Values: [][]byte{buf}}
+	encoded := message.Encode(nil)
+	payload := bytes.Repeat(encoded, repeats)
+	return &multiMessage{
+		singleMessage: message,
+		payload:       payload,
+	}, nil
+}
+
+func (m *multiMessage) Decode(_ []byte) error {
+	return trace.NotImplemented("Decode is not implemented for multiMessage")
+}
+
+func (m *multiMessage) Encode(dst []byte) []byte {
+	return append(dst, m.payload...)
+}
+
+func (m *multiMessage) Backend() {
+}
+
+var _ pgproto3.BackendMessage = (*multiMessage)(nil)
```

Comment:
If we ever update `pgproto3` and it ends up changing the `BackendMessage` interface, this particular line will help us diagnose the issue, because this kind of assert leads to a much better error message from the Go compiler.

Reply:
Suppose I there is a method `Backend()` which is missing. Now you are getting an error message in a random call site:

<img width="717" alt="image" src="https://github.com/gravitational/teleport/assets/158712/c785e395-6b64-495d-8ea3-94d1969b7777">

Clearly, this used to work in the past, but a signature has changed... which one? You need to go digging through the history and compare the versions to see the issue. You will get an answer, no doubt, but it will take a moment or two.

The usage may well be in some faraway file - or multiple, if the type is used multiple times. You will be getting dozens of errors in this way. You are using implicit usage of the type to assert its properties.

If instead you add a single line of assertion next to the type definition:

<img width="1008" alt="image" src="https://github.com/gravitational/teleport/assets/158712/2ae6862e-4551-491d-87ec-f5daa26bace6">

... then the intent is documented for posteriority, and we see clearly, which assertion is no longer valid. The file defining the type will no longer compile, and any errors will not be multiplied. Instead, you will get a single error right where the issue is.

I'd say this is somewhat analogous, if loosely, to integration tests vs unit tests. Both will point out if something doesn't work but in a rather different way.

Reply:
Hard to spot that by casually reading the code, though. When writing a type, you commonly want it to implement an interface. The method signatures will give you that interface, but explicitly calling it out makes it easier to read.